File: fraiseql-nodejs/src/security.ts
/**
 * Security module for FraiseQL Node.js
 *
 * Implements advanced authorization and security features including:
 * - Custom authorization rules with context variables
 * - Role-based access control (RBAC) with multiple strategies
 * - Attribute-based access control (ABAC)
 * - Reusable authorization policies
 * - Caching and audit logging
 */

/**
 * Defines how to match multiple roles in RBAC
 */
export enum RoleMatchStrategy {
  /** User must have at least one of the specified roles */
  ANY = 'any',

  /** User must have all of the specified roles */
  ALL = 'all',

  /** User must have exactly these roles, no more, no less */
  EXACTLY = 'exactly',
}

/**
 * Defines the type of authorization policy
 */
export enum AuthzPolicyType {
  /** Role-based access control */
  RBAC = 'rbac',

  /** Attribute-based access control */
  ABAC = 'abac',

  /** Custom rule expressions */
  CUSTOM = 'custom',

  /** Hybrid approach combining multiple methods */
  HYBRID = 'hybrid',
}

/**
 * Configuration for custom authorization rules
 */
export interface AuthorizeConfig {
  /** Authorization rule expression */
  rule?: string;

  /** Reference to a named policy */
  policy?: string;

  /** Description of what this rule protects */
  description?: string;

  /** Custom error message on denial */
  errorMessage?: string;

  /** Whether to apply hierarchically to child fields */
  recursive?: boolean;

  /** Operation-specific rules (read, create, update, delete) */
  operations?: string;

  /** Whether to cache authorization decisions */
  cacheable?: boolean;

  /** Cache duration in seconds */
  cacheDurationSeconds?: number;
}

/**
 * Configuration for role-based access control
 */
export interface RoleRequiredConfig {
  /** Required roles */
  roles?: string[];

  /** Role matching strategy (ANY, ALL, EXACTLY) */
  strategy?: RoleMatchStrategy;

  /** Whether roles form a hierarchy */
  hierarchy?: boolean;

  /** Description of the role requirement */
  description?: string;

  /** Custom error message on denial */
  errorMessage?: string;

  /** Operation-specific rules */
  operations?: string;

  /** Whether to inherit role requirements from parent types */
  inherit?: boolean;

  /** Whether to cache role validation results */
  cacheable?: boolean;

  /** Cache duration in seconds */
  cacheDurationSeconds?: number;
}

/**
 * Configuration for reusable authorization policies
 */
export interface AuthzPolicyConfig {
  /** Policy name */
  name: string;

  /** Policy description */
  description?: string;

  /** Authorization rule expression */
  rule?: string;

  /** Attribute conditions for ABAC policies */
  attributes?: string[];

  /** Policy type (RBAC, ABAC, CUSTOM, HYBRID) */
  type?: AuthzPolicyType;

  /** Whether to cache authorization decisions */
  cacheable?: boolean;

  /** Cache duration in seconds */
  cacheDurationSeconds?: number;

  /** Whether to apply recursively to nested types */
  recursive?: boolean;

  /** Operation-specific rules */
  operations?: string;

  /** Whether to log access decisions */
  auditLogging?: boolean;

  /** Custom error message */
  errorMessage?: string;
}

/**
 * Builder for custom authorization rules
 *
 * Example:
 * ```typescript
 * new AuthorizeBuilder()
 *   .rule("isOwner($context.userId, $field.ownerId)")
 *   .description("Ensures users can only access their own notes")
 *   .build();
 * ```
 */
export class AuthorizeBuilder {
  private config: AuthorizeConfig = {
    cacheable: true,
    cacheDurationSeconds: 300,
  };

  /**
   * Set the authorization rule expression
   */
  rule(rule: string): AuthorizeBuilder {
    this.config.rule = rule;
    return this;
  }

  /**
   * Set the reference to a named authorization policy
   */
  policy(policy: string): AuthorizeBuilder {
    this.config.policy = policy;
    return this;
  }

  /**
   * Set the description of what this rule protects
   */
  description(description: string): AuthorizeBuilder {
    this.config.description = description;
    return this;
  }

  /**
   * Set the custom error message
   */
  errorMessage(errorMessage: string): AuthorizeBuilder {
    this.config.errorMessage = errorMessage;
    return this;
  }

  /**
   * Set whether to apply rule hierarchically to child fields
   */
  recursive(recursive: boolean): AuthorizeBuilder {
    this.config.recursive = recursive;
    return this;
  }

  /**
   * Set which operations this rule applies to (read, create, update, delete)
   */
  operations(operations: string): AuthorizeBuilder {
    this.config.operations = operations;
    return this;
  }

  /**
   * Set whether to cache authorization decisions
   */
  cacheable(cacheable: boolean): AuthorizeBuilder {
    this.config.cacheable = cacheable;
    return this;
  }

  /**
   * Set the cache duration in seconds
   */
  cacheDurationSeconds(duration: number): AuthorizeBuilder {
    this.config.cacheDurationSeconds = duration;
    return this;
  }

  /**
   * Build the authorization configuration
   */
  build(): AuthorizeConfig {
    return { ...this.config };
  }
}

/**
 * Builder for role-based access control rules
 *
 * Example:
 * ```typescript
 * new RoleRequiredBuilder()
 *   .roles('manager', 'director')
 *   .strategy(RoleMatchStrategy.ANY)
 *   .description("Managers and directors can view salaries")
 *   .build();
 * ```
 */
export class RoleRequiredBuilder {
  private config: RoleRequiredConfig = {
    strategy: RoleMatchStrategy.ANY,
    inherit: true,
    cacheable: true,
    cacheDurationSeconds: 600,
  };

  /**
   * Set required roles (variadic for convenience)
   */
  roles(...roles: string[]): RoleRequiredBuilder {
    this.config.roles = roles;
    return this;
  }

  /**
   * Set required roles from an array
   */
  rolesArray(roles: string[]): RoleRequiredBuilder {
    this.config.roles = roles;
    return this;
  }

  /**
   * Set the role matching strategy
   */
  strategy(strategy: RoleMatchStrategy): RoleRequiredBuilder {
    this.config.strategy = strategy;
    return this;
  }

  /**
   * Set whether roles form a hierarchy
   */
  hierarchy(hierarchy: boolean): RoleRequiredBuilder {
    this.config.hierarchy = hierarchy;
    return this;
  }

  /**
   * Set the description of the role requirement
   */
  description(description: string): RoleRequiredBuilder {
    this.config.description = description;
    return this;
  }

  /**
   * Set the custom error message
   */
  errorMessage(errorMessage: string): RoleRequiredBuilder {
    this.config.errorMessage = errorMessage;
    return this;
  }

  /**
   * Set which operations this rule applies to
   */
  operations(operations: string): RoleRequiredBuilder {
    this.config.operations = operations;
    return this;
  }

  /**
   * Set whether to inherit role requirements from parent types
   */
  inherit(inherit: boolean): RoleRequiredBuilder {
    this.config.inherit = inherit;
    return this;
  }

  /**
   * Set whether to cache role validation results
   */
  cacheable(cacheable: boolean): RoleRequiredBuilder {
    this.config.cacheable = cacheable;
    return this;
  }

  /**
   * Set the cache duration in seconds
   */
  cacheDurationSeconds(duration: number): RoleRequiredBuilder {
    this.config.cacheDurationSeconds = duration;
    return this;
  }

  /**
   * Build the role configuration
   */
  build(): RoleRequiredConfig {
    return { ...this.config };
  }
}

/**
 * Builder for reusable authorization policies
 *
 * Example:
 * ```typescript
 * new AuthzPolicyBuilder('piiAccess')
 *   .type(AuthzPolicyType.RBAC)
 *   .rule("hasRole($context, 'data_manager') OR hasScope($context, 'read:pii')")
 *   .description("Access to Personally Identifiable Information")
 *   .build();
 * ```
 */
export class AuthzPolicyBuilder {
  private config: AuthzPolicyConfig;

  constructor(name: string) {
    this.config = {
      name,
      type: AuthzPolicyType.CUSTOM,
      cacheable: true,
      cacheDurationSeconds: 300,
      auditLogging: false,
    };
  }

  /**
   * Set the policy description
   */
  description(description: string): AuthzPolicyBuilder {
    this.config.description = description;
    return this;
  }

  /**
   * Set the authorization rule expression
   */
  rule(rule: string): AuthzPolicyBuilder {
    this.config.rule = rule;
    return this;
  }

  /**
   * Set attribute conditions for ABAC policies (variadic)
   */
  attributes(...attributes: string[]): AuthzPolicyBuilder {
    this.config.attributes = attributes;
    return this;
  }

  /**
   * Set attribute conditions from an array
   */
  attributesArray(attributes: string[]): AuthzPolicyBuilder {
    this.config.attributes = attributes;
    return this;
  }

  /**
   * Set the policy type
   */
  type(type: AuthzPolicyType): AuthzPolicyBuilder {
    this.config.type = type;
    return this;
  }

  /**
   * Set whether to cache authorization decisions
   */
  cacheable(cacheable: boolean): AuthzPolicyBuilder {
    this.config.cacheable = cacheable;
    return this;
  }

  /**
   * Set the cache duration in seconds
   */
  cacheDurationSeconds(duration: number): AuthzPolicyBuilder {
    this.config.cacheDurationSeconds = duration;
    return this;
  }

  /**
   * Set whether to apply recursively to nested types
   */
  recursive(recursive: boolean): AuthzPolicyBuilder {
    this.config.recursive = recursive;
    return this;
  }

  /**
   * Set which operations this policy applies to
   */
  operations(operations: string): AuthzPolicyBuilder {
    this.config.operations = operations;
    return this;
  }

  /**
   * Set whether to log access decisions
   */
  auditLogging(auditLogging: boolean): AuthzPolicyBuilder {
    this.config.auditLogging = auditLogging;
    return this;
  }

  /**
   * Set the custom error message
   */
  errorMessage(errorMessage: string): AuthzPolicyBuilder {
    this.config.errorMessage = errorMessage;
    return this;
  }

  /**
   * Build the authorization policy configuration
   */
  build(): AuthzPolicyConfig {
    return { ...this.config };
  }
}

/**
 * Decorator for custom authorization rules
 */
export function Authorize(config: AuthorizeConfig): MethodDecorator & ClassDecorator {
  return function <T extends Function>(target: T | any, propertyKey?: string | symbol): any {
    return target;
  };
}

/**
 * Decorator for role-based access control
 */
export function RoleRequired(config: RoleRequiredConfig): MethodDecorator & ClassDecorator {
  return function <T extends Function>(target: T | any, propertyKey?: string | symbol): any {
    return target;
  };
}

/**
 * Decorator for authorization policies
 */
export function AuthzPolicy(config: AuthzPolicyConfig): ClassDecorator & MethodDecorator {
  return function <T extends Function>(target: T | any, propertyKey?: string | symbol): any {
    return target;
  };
}

File: fraiseql-kotlin/src/main/kotlin/com/fraiseql/security/Security.kt
package com.fraiseql.security

/**
 * Sealed class for role matching strategies in RBAC
 */
sealed class RoleMatchStrategy(val value: String) {
    object Any : RoleMatchStrategy("any")
    object All : RoleMatchStrategy("all")
    object Exactly : RoleMatchStrategy("exactly")
}

/**
 * Sealed class for authorization policy types
 */
sealed class AuthzPolicyType(val value: String) {
    object Rbac : AuthzPolicyType("rbac")
    object Abac : AuthzPolicyType("abac")
    object Custom : AuthzPolicyType("custom")
    object Hybrid : AuthzPolicyType("hybrid")
}

/**
 * Configuration for custom authorization rules
 */
data class AuthorizeConfig(
    val rule: String = "",
    val policy: String = "",
    val description: String = "",
    val errorMessage: String = "",
    val recursive: Boolean = false,
    val operations: String = "",
    val cacheable: Boolean = true,
    val cacheDurationSeconds: Int = 300
) {
    fun toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        if (rule.isNotEmpty()) map["rule"] = rule
        if (policy.isNotEmpty()) map["policy"] = policy
        if (description.isNotEmpty()) map["description"] = description
        if (errorMessage.isNotEmpty()) map["error_message"] = errorMessage
        if (recursive) map["recursive"] = recursive
        if (operations.isNotEmpty()) map["operations"] = operations
        if (cacheable) {
            map["cacheable"] = cacheable
            map["cache_duration_seconds"] = cacheDurationSeconds
        }
        return map
    }
}

/**
 * Configuration for role-based access control
 */
data class RoleRequiredConfig(
    val roles: List<String> = emptyList(),
    val strategy: RoleMatchStrategy = RoleMatchStrategy.Any,
    val hierarchy: Boolean = false,
    val description: String = "",
    val errorMessage: String = "",
    val operations: String = "",
    val inherit: Boolean = true,
    val cacheable: Boolean = true,
    val cacheDurationSeconds: Int = 600
) {
    fun toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        if (roles.isNotEmpty()) map["roles"] = roles
        if (strategy != RoleMatchStrategy.Any) map["strategy"] = strategy.value
        if (hierarchy) map["hierarchy"] = hierarchy
        if (description.isNotEmpty()) map["description"] = description
        if (errorMessage.isNotEmpty()) map["error_message"] = errorMessage
        if (operations.isNotEmpty()) map["operations"] = operations
        if (!inherit) map["inherit"] = inherit
        if (cacheable) {
            map["cacheable"] = cacheable
            map["cache_duration_seconds"] = cacheDurationSeconds
        }
        return map
    }
}

/**
 * Configuration for reusable authorization policies
 */
data class AuthzPolicyConfig(
    val name: String,
    val description: String = "",
    val rule: String = "",
    val attributes: List<String> = emptyList(),
    val type: AuthzPolicyType = AuthzPolicyType.Custom,
    val cacheable: Boolean = true,
    val cacheDurationSeconds: Int = 300,
    val recursive: Boolean = false,
    val operations: String = "",
    val auditLogging: Boolean = false,
    val errorMessage: String = ""
) {
    fun toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>("name" to name)
        if (description.isNotEmpty()) map["description"] = description
        if (rule.isNotEmpty()) map["rule"] = rule
        if (attributes.isNotEmpty()) map["attributes"] = attributes
        if (type != AuthzPolicyType.Custom) map["type"] = type.value
        if (cacheable) {
            map["cacheable"] = cacheable
            map["cache_duration_seconds"] = cacheDurationSeconds
        }
        if (recursive) map["recursive"] = recursive
        if (operations.isNotEmpty()) map["operations"] = operations
        if (auditLogging) map["audit_logging"] = auditLogging
        if (errorMessage.isNotEmpty()) map["error_message"] = errorMessage
        return map
    }
}

/**
 * Builder for custom authorization rules
 *
 * Example:
 * ```kotlin
 * AuthorizeBuilder()
 *   .rule("isOwner(\$context.userId, \$field.ownerId)")
 *   .description("Ensures users can only access their own notes")
 *   .build()
 * ```
 */
class AuthorizeBuilder {
    private var rule: String = ""
    private var policy: String = ""
    private var description: String = ""
    private var errorMessage: String = ""
    private var recursive: Boolean = false
    private var operations: String = ""
    private var cacheable: Boolean = true
    private var cacheDurationSeconds: Int = 300

    fun rule(rule: String) = apply { this.rule = rule }
    fun policy(policy: String) = apply { this.policy = policy }
    fun description(description: String) = apply { this.description = description }
    fun errorMessage(errorMessage: String) = apply { this.errorMessage = errorMessage }
    fun recursive(recursive: Boolean) = apply { this.recursive = recursive }
    fun operations(operations: String) = apply { this.operations = operations }
    fun cacheable(cacheable: Boolean) = apply { this.cacheable = cacheable }
    fun cacheDurationSeconds(duration: Int) = apply { this.cacheDurationSeconds = duration }

    fun build() = AuthorizeConfig(
        rule = rule,
        policy = policy,
        description = description,
        errorMessage = errorMessage,
        recursive = recursive,
        operations = operations,
        cacheable = cacheable,
        cacheDurationSeconds = cacheDurationSeconds
    )
}

/**
 * Builder for role-based access control rules
 *
 * Example:
 * ```kotlin
 * RoleRequiredBuilder()
 *   .roles("manager", "director")
 *   .strategy(RoleMatchStrategy.Any)
 *   .description("Managers and directors can view salaries")
 *   .build()
 * ```
 */
class RoleRequiredBuilder {
    private var roles: List<String> = emptyList()
    private var strategy: RoleMatchStrategy = RoleMatchStrategy.Any
    private var hierarchy: Boolean = false
    private var description: String = ""
    private var errorMessage: String = ""
    private var operations: String = ""
    private var inherit: Boolean = true
    private var cacheable: Boolean = true
    private var cacheDurationSeconds: Int = 600

    fun roles(vararg roles: String) = apply { this.roles = roles.toList() }
    fun rolesArray(roles: List<String>) = apply { this.roles = roles }
    fun strategy(strategy: RoleMatchStrategy) = apply { this.strategy = strategy }
    fun hierarchy(hierarchy: Boolean) = apply { this.hierarchy = hierarchy }
    fun description(description: String) = apply { this.description = description }
    fun errorMessage(errorMessage: String) = apply { this.errorMessage = errorMessage }
    fun operations(operations: String) = apply { this.operations = operations }
    fun inherit(inherit: Boolean) = apply { this.inherit = inherit }
    fun cacheable(cacheable: Boolean) = apply { this.cacheable = cacheable }
    fun cacheDurationSeconds(duration: Int) = apply { this.cacheDurationSeconds = duration }

    fun build() = RoleRequiredConfig(
        roles = roles,
        strategy = strategy,
        hierarchy = hierarchy,
        description = description,
        errorMessage = errorMessage,
        operations = operations,
        inherit = inherit,
        cacheable = cacheable,
        cacheDurationSeconds = cacheDurationSeconds
    )
}

/**
 * Builder for reusable authorization policies
 *
 * Example:
 * ```kotlin
 * AuthzPolicyBuilder("piiAccess")
 *   .type(AuthzPolicyType.Rbac)
 *   .rule("hasRole(\$context, 'data_manager') OR hasScope(\$context, 'read:pii')")
 *   .description("Access to Personally Identifiable Information")
 *   .build()
 * ```
 */
class AuthzPolicyBuilder(private val name: String) {
    private var description: String = ""
    private var rule: String = ""
    private var attributes: List<String> = emptyList()
    private var type: AuthzPolicyType = AuthzPolicyType.Custom
    private var cacheable: Boolean = true
    private var cacheDurationSeconds: Int = 300
    private var recursive: Boolean = false
    private var operations: String = ""
    private var auditLogging: Boolean = false
    private var errorMessage: String = ""

    fun description(description: String) = apply { this.description = description }
    fun rule(rule: String) = apply { this.rule = rule }
    fun attributes(vararg attributes: String) = apply { this.attributes = attributes.toList() }
    fun attributesArray(attributes: List<String>) = apply { this.attributes = attributes }
    fun type(type: AuthzPolicyType) = apply { this.type = type }
    fun cacheable(cacheable: Boolean) = apply { this.cacheable = cacheable }
    fun cacheDurationSeconds(duration: Int) = apply { this.cacheDurationSeconds = duration }
    fun recursive(recursive: Boolean) = apply { this.recursive = recursive }
    fun operations(operations: String) = apply { this.operations = operations }
    fun auditLogging(auditLogging: Boolean) = apply { this.auditLogging = auditLogging }
    fun errorMessage(errorMessage: String) = apply { this.errorMessage = errorMessage }

    fun build() = AuthzPolicyConfig(
        name = name,
        description = description,
        rule = rule,
        attributes = attributes,
        type = type,
        cacheable = cacheable,
        cacheDurationSeconds = cacheDurationSeconds,
        recursive = recursive,
        operations = operations,
        auditLogging = auditLogging,
        errorMessage = errorMessage
    )
}

/**
 * Annotation for custom authorization rules
 */
@Target(AnnotationTarget.CLASS, AnnotationTarget.PROPERTY)
@Retention(AnnotationRetention.RUNTIME)
annotation class Authorize(
    val rule: String = "",
    val policy: String = "",
    val description: String = "",
    val errorMessage: String = "",
    val recursive: Boolean = false,
    val operations: String = "",
    val cacheable: Boolean = true,
    val cacheDurationSeconds: Int = 300
)

/**
 * Annotation for role-based access control
 */
@Target(AnnotationTarget.CLASS, AnnotationTarget.PROPERTY)
@Retention(AnnotationRetention.RUNTIME)
annotation class RoleRequired(
    val roles: Array<String> = [],
    val strategy: String = "any",
    val hierarchy: Boolean = false,
    val description: String = "",
    val errorMessage: String = "",
    val operations: String = "",
    val inherit: Boolean = true,
    val cacheable: Boolean = true,
    val cacheDurationSeconds: Int = 600
)

/**
 * Annotation for authorization policies
 */
@Target(AnnotationTarget.CLASS, AnnotationTarget.PROPERTY)
@Retention(AnnotationRetention.RUNTIME)
annotation class AuthzPolicy(
    val name: String,
    val description: String = "",
    val rule: String = "",
    val attributes: Array<String> = [],
    val type: String = "custom",
    val cacheable: Boolean = true,
    val cacheDurationSeconds: Int = 300,
    val recursive: Boolean = false,
    val operations: String = "",
    val auditLogging: Boolean = false,
    val errorMessage: String = ""
)
